Homework 3/Homeworks MS/Search/Search.c

```diff
@@ -0,0 +1,179 @@
+#include <stdio.h>
+#include <time.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+int* makeAnArray(int lengthOfArray)
+{
+    return calloc(lengthOfArray, sizeof(int));
+}
+ 
+void makeAnArrayOfRandomNumbers(int array[], int lengthOfArray)
+{
+    srand(time(NULL));
+    for (int i = 0; i < lengthOfArray; ++i)
+    {
+        array[i] = -100 + rand() % 200;
+    }
+}
+
+void qSort(int array[], int startIndex, int stopIndex)
+{
+    if (stopIndex - startIndex == 0)
+    {
+        return;
+    }
+    if (stopIndex - startIndex == 1)
+    {
+        if (array[startIndex] > array[stopIndex])
+        {
+            int const buffer = array[startIndex];
+            array[startIndex] = array[stopIndex];
+            array[stopIndex] = buffer;
+        }
+        return;
+    }
+    int i = startIndex + 1;
+    while (array[i - 1] == array[i] && i <= stopIndex)
+    {
+        ++i;
+    }
+    if (i == stopIndex + 1 || array[i] == array[i - 1])
+    {
+        return;
+    }
+    int supportElement = 0;
+    if (array[i] > array[i - 1])
+    {
+        supportElement = array[i];
+    }
+    else
+    {
+        supportElement = array[i - 1];
+    }
```

Comment:
Магия тернарного оператора делает код короче и чище:
```suggestion
    int supportElement = array[i] > array[i - 1] ? array[i] : array[i - 1];
```
Хотя есть библиотечный макрос max(), который делает то же самое, я бы использовал его (но Вы вправе про него не знать, а вот тернарный оператор знать надо).

Reply:
И тут тоже, выбор опорного элемента стоит в отдельную функцию вынести, он достаточно сложный, чтобы этого заслуживать. Андрей Николаевич любит говорить, что если функция не влезает на экран хоть чуть-чуть, вероятность ошибки в ней гораздо выше, чем если влезает. Поэтому делите большие и страшные функции на маленькие и нестрашные.